File: mandelbrot/ruby/mandelbrot.rb
#!/usr/bin/ruby

require 'complex'
require 'rubygems'
require 'chunky_png'

class Mandelbrot
  def initialize cap
    @cap = cap
  end

  # given a complex number, return iterations to escape, up to cap,
  # or zero if no escape occurs
  def escape c
    z = Complex(0, 0)
    for it in 1 .. @cap
      z = z ** 2 + c
      break if z.abs2 > 4.0     # 2.0 * 2.0
    end
    (it == @cap) ? 0 : it
  end
end

class MandelImage < Mandelbrot
  DENSITY_FACTOR = 16
  def initialize width, height, cap
    super cap
    @width = width
    @height = height
    @scalec = 0xffffff / cap * DENSITY_FACTOR
    zoom                        # start at default zoom
  end

  def zoom xmin = -2.5, xmax = 1.0, ymin = -1.0, ymax = 1.0
    @xmin = xmin
    @xmax = xmax
    @ymin = ymin
    @ymax = ymax
    @xstep = (xmax - xmin).abs / @width
    @ystep = (ymax - ymin).abs / @height
    self
  end

  def image
    img = ChunkyPNG::Image.new(@width, @height, ChunkyPNG::Color::TRANSPARENT)

    (0..(@height-1)).each do |y|
      (0..(@width-1)).each do |x|
        c = Complex scalex(x), scaley(y)
        it = escape c
        img[x, y] = color it
      end
    end 
    img
  end

private

  def scalex n
    @xmin + (n * @xstep)
  end

  def scaley n
    @ymin + (n * @ystep)
  end

  def color n
    return ChunkyPNG::Color::BLACK if n == 0
    n = n * @scalec
    
    r = n & 0xff
    g = (n & 0xff00) >> 8
    b = (n & 0xff0000) >> 16
    ChunkyPNG::Color::rgb(r,g,b)
  end  
end

MandelImage.new(1280, 800, 1000).image.save("mandelbrot-ruby.png")
MandelImage.new(1280, 800, 1000).zoom(-0.5, 0.5, 0.0, 0.75).image.save("mandelzoom1-ruby.png")

File: mandelbrot/python/mandelbrot.py
#!/opt/csw/bin/python

import Image, ImageDraw

class Mandelbrot:
  def __init__(self, cap):
    self.cap = cap

  def escape(self, c):
    z = complex(0, 0)
    for it in range(self.cap + 1):
      if abs(z) > 2.0:
        return it
      z = (z ** 2) + c
    return 0

class MandelImage(Mandelbrot):
  def __init__(self, width, height, cap):
    # super(MandelImage).__init__(cap)
    self.cap = cap
    self.width = width
    self.height = height
    self.scalec = 0xffffff / cap
    self.zoom()
    self.density_factor = 16

  def zoom(self, xmin=-2.5, xmax=1.0, ymin=-1.0, ymax=1.0):
    self.xmin = xmin
    self.xmax = xmax
    self.ymin = ymin
    self.ymax = ymax
    self.xstep = abs(xmax - xmin) / self.width
    self.ystep = abs(ymax - ymin) / self.height
    return self

  def image(self):
    img = Image.new("RGB", (self.width, self.height))
    draw = ImageDraw.Draw(img)
    for y in range(self.height):
      for x in range(self.width):
        c = complex(self.scalex(x), (self.scaley(y)))
        it = self.escape(c)
        draw.point((x, y), self.color(it))
    return img
  
  def scalex(self, n):
    return self.xmin + (n * self.xstep)

  def scaley(self, n):
    return self.ymin + (n * self.ystep)

  def color(self, n):
    if n == 0:
      return (0,0,0) 
    n = (n * self.scalec * self.density_factor) % 0xffffff
    r =  (n & 0xff)
    g =  ((n & 0xff00) >> 8)
    b =  ((n & 0xff0000) >> 16)
    return (r,g,b)

MandelImage(1280,800,1000).image().save("mandelbrot-python.png")
MandelImage(1280,800,1000).zoom(-0.5, 0.5, 0, 0.75).image().save("mandelzoom1-python.png")
